File: 3729-unit-conversion-i/3729-unit-conversion-i.py
class Solution(object):
    def baseUnitConversions(self, conversions):
        """
        :type conversions: List[List[int]]
        :rtype: List[int]
        """
        n = len(conversions)
        mod = 1000000007
        adj = [[] for _ in range(n + 1)]
        for conversion in conversions:
            f = conversion[0]
            t = conversion[1]
            r = conversion[2]
            adj[f].append((t, r))
        #print(adj)
        q = deque()
        q.append(0)
        
        ans = [-1 for _ in range(n + 1)]
        ans[0] = 1
        
        while q:
            top = q.popleft()
            for edge in adj[top]:
                if ans[edge[0]] == -1:
                    ans[edge[0]] = (ans[top] * edge[1]) % mod
                    q.append(edge[0])
        return ans

File: 3729-unit-conversion-i/3729-unit-conversion-i.go
func baseUnitConversions(conversions [][]int) []int {
    mod := 1000_000_000 + 7

    n := len(conversions) + 1

    ans := make([]int,n)
    for i := 0; i < n; i++ {
        ans[i] = -1
    }
    
    egs := make([][][2]int,n)
    for i := range conversions {
        u, v, w := conversions[i][0], conversions[i][1], conversions[i][2]
        egs[u] = append(egs[u],[2]int{v, w})
    }

    qu := []int {0}
    ans[0] = 1
    for len(qu) > 0 {
        fr := qu[0]
        qu = qu[1: ]
        for i := range egs[fr] {
            t,w := egs[fr][i][0],egs[fr][i][1]
            if ans[t] != -1 {
                continue
            }
            ans[t] = ans[fr] * w % mod
            qu = append(qu,t)
        }
    }
    
    return ans
}